File: src/core/Stocks/Events.cs
using System;
using core.Shared;

namespace core.Stocks
{
    internal class StockDeleted : AggregateEvent
    {
        public StockDeleted(Guid id, Guid aggregateId, DateTimeOffset when) : base(id, aggregateId, when)
        {
        }
    }

    public class StockPurchased : 
        AggregateEvent,
        IStockTransaction
    {
        public StockPurchased(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string ticker,
            decimal numberOfShares,
            decimal price,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
            NumberOfShares = numberOfShares;
            Price = price;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal NumberOfShares { get; }
        public decimal Price { get; }
        public string Notes { get; }
    }

    public class StockPurchased_v2 : 
        AggregateEvent,
        IStockTransactionWithStopPrice
    {
        public StockPurchased_v2(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string ticker,
            decimal numberOfShares,
            decimal price,
            string notes,
            decimal? stopPrice)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
            NumberOfShares = numberOfShares;
            Price = price;
            Notes = notes;
            StopPrice = stopPrice;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal NumberOfShares { get; }
        public decimal Price { get; }
        public string Notes { get; }
        public decimal? StopPrice { get; }
    }

    public class StockSold :
        AggregateEvent,
        IStockTransaction
    {
        public StockSold(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string ticker,
            decimal numberOfShares,
            decimal price,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
            NumberOfShares = numberOfShares;
            Price = price;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal NumberOfShares { get; }
        public decimal Price { get; }
        public string Notes { get; }
    }

    internal class StockTransactionDeleted : AggregateEvent
    {
        public StockTransactionDeleted(
            Guid id,
            Guid aggregateId,
            Guid transactionId,
            DateTimeOffset when) : base(id, aggregateId, when)
        {
            TransactionId = transactionId;
        }

        public Guid TransactionId { get; }
    }

    [Obsolete("Use position labels instead")]
    internal class StockCategoryChanged : AggregateEvent
    {
        public StockCategoryChanged(Guid id, Guid aggregateId, string category, DateTimeOffset when) : base(id, aggregateId, when)
        {
            Category =category;
        }

        public string Category { get; }
    }

    internal class TickerObtained : AggregateEvent
    {
        public TickerObtained(Guid id, Guid aggregateId, DateTimeOffset when, string ticker, Guid userId) : base(id, aggregateId, when)
        {
            Ticker = ticker;
            UserId = userId;
        }

        public string Ticker { get; }
        public Guid UserId { get; }
    }

    public class StopPriceSet : AggregateEvent
    {
        public StopPriceSet(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, string ticker, decimal stopPrice) : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
            StopPrice = stopPrice;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal StopPrice { get; }
    }

    public class NotesAdded : AggregateEvent
    {
        public NotesAdded(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, int positionId, string notes) : base(id, aggregateId, when)
        {
            UserId = userId;
            PositionId = positionId;
            Notes = notes;
        }

        public Guid UserId { get; }
        public int PositionId { get; }
        public string Notes { get; }
    }

    internal class RiskAmountSet : AggregateEvent
    {
        public RiskAmountSet(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, string ticker, decimal riskAmount) : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
            RiskAmount = riskAmount;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal RiskAmount { get; }
    }

    public class StopDeleted : AggregateEvent
    {
        public StopDeleted(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, string ticker) : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
    }

    public class TradeGradeAssigned : AggregateEvent
    {
        public TradeGradeAssigned(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, string grade, string note, int positionId) : base(id, aggregateId, when)
        {
            UserId = userId;
            Grade = grade;
            Note = note;
            PositionId = positionId;
        }

        public Guid UserId { get; }
        public int PositionId { get; }
        public string Grade { get; }
        public string Note { get; }
    }

    public class PositionDeleted : AggregateEvent
    {
        public PositionDeleted(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, int positionId) : base(id, aggregateId, when)
        {
            PositionId = positionId;
            UserId = userId;
        }
        public int PositionId { get; }
        public Guid UserId { get; }
    }

    public class PositionLabelSet : AggregateEvent
    {
        public PositionLabelSet(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, int positionId, string key, string value) :
            base(id, aggregateId, when)
        {
            UserId = userId;
            PositionId = positionId;
            Key = key;
            Value = value;
        }

        public Guid UserId { get; }
        public int PositionId { get; }
        public string Key { get; }
        public string Value { get; }
    }

    public class PositionLabelDeleted : AggregateEvent
    {
        public PositionLabelDeleted(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, int positionId, string key) :
            base(id, aggregateId, when)
        {
            UserId = userId;
            PositionId = positionId;
            Key = key;
        }

        public Guid UserId { get; }
        public int PositionId { get; }
        public string Key { get; }
    }

    public class PositionRiskAmountSet : AggregateEvent
    {
        public PositionRiskAmountSet(Guid id, Guid aggregateId, DateTimeOffset when, Guid userId, int positionId, decimal riskAmount) :
            base(id, aggregateId, when)
        {
            UserId = userId;
            PositionId = positionId;
            RiskAmount = riskAmount;
        }

        public Guid UserId { get; }
        public int PositionId { get; }
        public decimal RiskAmount { get; }
    }
}

File: src/core2.fs/Stocks/Events.fs
namespace core.Stocks

open System
open core.Shared

type IStockTransaction =
    abstract NumberOfShares: decimal
    abstract Price: decimal
    abstract Id: Guid
    abstract When: DateTimeOffset
    abstract Notes: string

type IStockTransactionWithStopPrice =
    inherit IStockTransaction
    abstract StopPrice: decimal option

[<AllowNullLiteral>]
type internal StockDeleted(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)

type StockPurchased(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string, numberOfShares: decimal, price: decimal, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val NumberOfShares = numberOfShares with get
    member val Price = price with get
    member val Notes = notes with get
    interface IStockTransaction with
        member this.NumberOfShares = this.NumberOfShares
        member this.Price = this.Price
        member this.Id = this.Id
        member this.When = this.When
        member this.Notes = this.Notes

type StockPurchased_v2(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string, numberOfShares: decimal, price: decimal, notes: string, stopPrice: decimal option) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val NumberOfShares = numberOfShares with get
    member val Price = price with get
    member val Notes = notes with get
    member val StopPrice = stopPrice with get
    interface IStockTransactionWithStopPrice with
        member this.NumberOfShares = this.NumberOfShares
        member this.Price = this.Price
        member this.Id = this.Id
        member this.When = this.When
        member this.Notes = this.Notes
        member this.StopPrice = this.StopPrice

type StockSold(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string, numberOfShares: decimal, price: decimal, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val NumberOfShares = numberOfShares with get
    member val Price = price with get
    member val Notes = notes with get
    interface IStockTransaction with
        member this.NumberOfShares = this.NumberOfShares
        member this.Price = this.Price
        member this.Id = this.Id
        member this.When = this.When
        member this.Notes = this.Notes

[<AllowNullLiteral>]
type internal StockTransactionDeleted(id: Guid, aggregateId: Guid, transactionId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val TransactionId = transactionId with get

[<Obsolete("Use position labels instead")>]
[<AllowNullLiteral>]
type internal StockCategoryChanged(id: Guid, aggregateId: Guid, category: string, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Category = category with get

[<AllowNullLiteral>]
type internal TickerObtained(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, ticker: string, userId: Guid) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Ticker = ticker with get
    member val UserId = userId with get

[<AllowNullLiteral>]
type StopPriceSet(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string, stopPrice: decimal) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val StopPrice = stopPrice with get

[<AllowNullLiteral>]
type NotesAdded(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, positionId: int, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val PositionId = positionId with get
    member val Notes = notes with get

[<AllowNullLiteral>]
type internal RiskAmountSet(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string, riskAmount: decimal) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val RiskAmount = riskAmount with get

[<AllowNullLiteral>]
type StopDeleted(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get

[<AllowNullLiteral>]
type TradeGradeAssigned(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, grade: string, note: string, positionId: int) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val PositionId = positionId with get
    member val Grade = grade with get
    member val Note = note with get

[<AllowNullLiteral>]
type PositionDeleted(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, positionId: int) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val PositionId = positionId with get
    member val UserId = userId with get

[<AllowNullLiteral>]
type PositionLabelSet(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, positionId: int, key: string, value: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val PositionId = positionId with get
    member val Key = key with get
    member val Value = value with get

[<AllowNullLiteral>]
type PositionLabelDeleted(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, positionId: int, key: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val PositionId = positionId with get
    member val Key = key with get

[<AllowNullLiteral>]
type PositionRiskAmountSet(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, positionId: int, riskAmount: decimal) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val PositionId = positionId with get
    member val RiskAmount = riskAmount with get
